source/opt/def_use_manager.cpp

```diff
@@ -42,48 +42,76 @@ void DefUseManager::AnalyzeDefUse(ir::Module* module) {
+  return id_to_def_.at(id);
 }
 
-std::list<DefUseManager::InstLocPair>* DefUseManager::GetUses(uint32_t id) {
+UseList* DefUseManager::GetUses(uint32_t id) {
   if (id_to_uses_.count(id) == 0) return nullptr;
   return &id_to_uses_.at(id);
 }
 
 bool DefUseManager::KillDef(uint32_t id) {
-  if (id_to_defs_.count(id) == 0) return false;
-  id_to_defs_[id]->ToNop();
-  id_to_defs_.erase(id);
+  if (id_to_def_.count(id) == 0) return false;
+
+  // Go through all ids usd by this instruction, remove this instruction's uses
+  // of them.
+  for (const auto use_id : id_to_params_[id]) {
+    if (id_to_uses_.count(use_id) == 0) continue;
+    auto& uses = id_to_uses_[use_id];
+    for (auto it = uses.begin(); it != uses.end();) {
+      if (it->inst->result_id() == id) {
+        it = uses.erase(it);
+      } else {
+        ++it;
+      }
+    }
+    if (uses.empty()) id_to_uses_.erase(use_id);
+  }
+  id_to_params_.erase(id);
+
+  id_to_uses_.erase(id);  // Remove all uses of this id.
+  // This must happen at the last since we use information inside the instuction
+  // in the above.
+  id_to_def_[id]->ToNop();
+  id_to_def_.erase(id);
   return true;
 }
 
 bool DefUseManager::ReplaceAllUsesWith(uint32_t before, uint32_t after) {
+  if (before == after) return false;
   if (id_to_uses_.count(before) == 0) return false;
 
   for (auto it = id_to_uses_[before].cbegin(); it != id_to_uses_[before].cend();
        ++it) {
-    auto* inst = it->first;
-    auto loc = it->second;
     // Make the modification in the instruction.
-    inst->SetOperand(loc, {after});
+    it->inst->SetOperand(it->operand_index, {after});
     // Register the use of |after| id into id_to_uses_.
     // TODO(antiagainst): de-duplication.
-    id_to_uses_[after].emplace_back(inst, loc);
+    id_to_uses_[after].push_back({it->inst, it->operand_index});
   }
   id_to_uses_.erase(before);
   return true;
 }
 
 void DefUseManager::AnalyzeInstDefUse(ir::Instruction* inst) {
-  if (inst->result_id() != 0) id_to_defs_[inst->result_id()] = inst;
+  const uint32_t def_id = inst->result_id();
+  if (def_id != 0) id_to_def_[def_id] = inst;
 
   for (uint32_t i = 0; i < inst->NumOperands(); ++i) {
-    const auto type = inst->GetOperand(i).type;
-    if (type == SPV_OPERAND_TYPE_ID || type == SPV_OPERAND_TYPE_TYPE_ID ||
-        type == SPV_OPERAND_TYPE_MEMORY_SEMANTICS_ID ||
-        type == SPV_OPERAND_TYPE_SCOPE_ID) {
-      id_to_uses_[inst->GetSingleWordOperand(i)].emplace_back(inst, i);
+    switch (inst->GetOperand(i).type) {
+      // For any id type but result id type
+      case SPV_OPERAND_TYPE_ID:
+      case SPV_OPERAND_TYPE_TYPE_ID:
+      case SPV_OPERAND_TYPE_MEMORY_SEMANTICS_ID:
+      case SPV_OPERAND_TYPE_SCOPE_ID: {
+        uint32_t use_id = inst->GetSingleWordOperand(i);
+        // use_id is used by the instruction generating def_id.
+        id_to_uses_[use_id].push_back({inst, i});
+        id_to_params_[def_id].push_back(use_id);
```

Comment:
What if def_id is 0?  This will happen if the current instruction is something like OpDecorate (which does not return a result id).
Seem wrong.